First draft of HyP3 description section (#17)
ASFHyP3/hyp3-paper: paper/paper.md

TO DO

# HyP3 Description

## Development History

> HyP3 grew out of a need for a standardized pipeline to perform image processing workflows at the Alaska Satellite Facility (ASF) Distributed Active Archive Center (DAAC). HyP3 was initially created in 2017 by a group of undergraduate University of Alaska Fairbanks students interning at ASF. These students created a prototype version of HyP3 that was designed to handle workflows that ran on-premise and in the AWS cloud. This prototype was iterated upon for several more years, and by 2019 it was capable of performing many SAR-specific image processing workflows.

Reviewer: I decided to add this part of https://github.com/ASFHyP3/hyp3-paper/pull/19 as a suggestion here, in case that PR gets rejected, as this is the only really important part.
```suggestion
HyP3 grew out of a need for a standardized pipeline to perform image processing workflows at the Alaska Satellite Facility (ASF) Distributed Active Archive Center (DAAC). HyP3 was initially created in 2016 by a group of undergraduate students interning at ASF. These students created a prototype version of HyP3 that was designed to handle workflows that ran on-premise and in the AWS cloud. This prototype was iterated upon for several more years, and by 2019 it was capable of performing many SAR-specific image processing workflows.
```

By 2019 it was clear that re-designing HyP3 to take advantage of evolving technologies (namely the new AWS Batch service, infrastructure-as-code, and containerized workflows) would greatly improve its utility. Thus, in 2019 HyP3 was re-designed to run entirely within the AWS cloud, with application-specific processing workflows encapsulated within Docker containers. The re-design was completed in 2020, and on-demand SAR image processing services that utilized HyP3 were made available to the public at no cost via an API, Python SDK, and ASF's Data Search Vertex website.

> The first redesigned on-demand workflow made publicly available was Radiometric Terrain Correction (RTC) in [month year], followed by InSAR in [month year]. Since that time, new image processing services have been added to HyP3 including a glacier velocity tracking tool, and flood inundation mapping workflows. The user base grew to XX by [date], with users producing XX image products on average every month. Since 2021, a range of users have recognized the utility of HyP3. HyP3 was deployed within NASA's Earthdata Cloud in 2022, and many other deployments of HyP3 now exist at other institutions (see the Ongoing Projects section for details).

Reviewer 2: ```suggestion
The first re-designed on-demand workflow made publicly available was Radiometric Terrain Correction (RTC) in [month year], followed by InSAR in [month year]. Since that time, new image processing services have been added to HyP3 including a glacier velocity tracking tool, and flood inundation mapping workflows. The user base grew to XX by [date], with users producing XX image products on average every month. Since 2021, a range of users have recognized the utility of HyP3. HyP3 was deployed within NASA's Earthdata Cloud in 2022, and many other deployments of HyP3 now exist at other institutions (see the Ongoing Projects section for details).
```

Reviewer 2: Make hyphenation consistent with previous usage

## HyP3 Architecture

> HyP3 is a cloud processing pipeline that is designed with a serverless architecture and to follow AWS architecture best practices. It also strives to make a clear separation between the plugins that perform the image processing and the pipeline in which this task is performed. See Figure FF for the AWS architecture diagram.

Reviewer 2: ```suggestion
HyP3 is a cloud processing pipeline that is designed with a serverless architecture and follows the AWS Well-Architectured Framework. It also strives to make a clear separation between the plugins that perform the image processing and the pipeline in which this task is performed. See Figure FF for the AWS architecture diagram.
```

Reviewer 2: I felt like 'AWS best practices' felt a big imprecise. Did we use the [AWS Well-Architectured Framework](https://docs.aws.amazon.com/wellarchitected/latest/framework/welcome.html)? What exactly is meant by best practices if that's not what we used?

Reviewer: I doubt that the team followed any specific guidelines such as those, but I also agree the original wording is vague. I might suggest just removing that bit.

> HyP3 is a cloud processing pipeline that is designed with a serverless architecture and follows the AWS Well-Architected Framework. It also strives to make a clear separation between the plugins that perform the image processing and the pipeline in which this task is performed. See Figure FF for the AWS architecture diagram.

Reviewer: I've never heard of the AWS Well-Architected Framework so I doubt we intentionally followed it for the design of HyP3, but happy to be corrected.
```suggestion
HyP3 is a cloud processing pipeline with a serverless architecture. The design makes a clear distinction between HyP3 itself and the image processing plugins. See Figure FF for the AWS architecture diagram.
```

Reviewer 2: Good point, Jake. I put this the AWS Framework here because it seemed more specific than "best practices," but maybe just removing that comment would be the best move. We can talk about specific best practices we emphasized if we'd like, but otherwise it's a big vague.

Author: You hear about the well-architected framework A LOT when you take AWS certification trainings. I've reached out to Andrew on MM to get his take.

Users can interact with HyP3 via three access methods: a graphical interface that is integrated within ASF's data search platform (Vertex), a Python software development kit (SDK), or directly via an application program interface (API). The API follows the OpenAPI specification and provides GET/POST requests for submitting job requests to HyP3, checking user information, and for checking the status of previously requested jobs. Both the SDK and graphical interface translate their inputs into API calls. The API is maintained by an API Gateway (Figure FF), and all submitted jobs are stored within a DynamoDB NoSQL database.

> An EventBridge scheduled event is set to occur every 60 seconds, which triggers a Lambda function that checks the DynamoDB jobs table for newly submitted jobs and starts an AWS Step Function execution for each new job. Similar to what is possible in Apache Airflow, Step Functions defines a state machine that can be used to execute the processing workflow for a particular job.

Reviewer 2: ```suggestion
A scheduled EventBridge event triggers a Lambda function every 60 seconds to check the DynamoDB jobs table for newly submitted jobs and start an AWS Step Function execution for each new job. Similar to what is possible in Apache Airflow, Step Functions defines a state machine that can be used to execute the processing workflow for a particular job.
```

Reviewer 2: Overall, there's a lot of passive voice in this draft. Is that how we want to write? If so, totally ignore the first comment. 
For the second comment, the pluralization of the second 'Step Functions' was kind of jarring to read. It's a bit redundant, but making sure it's clear that Step Functions is the product name, that might make it easier to read?   This might be another totally personal preference. 
```suggestion
An EventBridge scheduled event is set to occur every 60 seconds, which triggers a Lambda function that checks the DynamoDB jobs table for newly submitted jobs and starts an AWS Step Function execution for each new job. Similar to what is possible in Apache Airflow, AWS Step Functions defines a state machine that can be used to execute the processing workflow for a particular job.
```

Author: I tend to prefer to prefer active voice, but sometimes default to passive out of habit.

Reviewer: I agree the wording around Step Functions should change. I gave up trying to figure out the best wording because I think we're just focusing on factual correctness and content completeness for the first draft, and we'll clean up the language later.

> The workflow begins by inspecting the job definition for the job type. This job type will be used to determine which image processing workflow is run and what computing resources the job needs. With this setup, a HyP3 deployment can contain an arbitrary number of different image processing workflows. Once the memory and compute resources requirements are determined, the workflow submits the job processing via AWS Batch, which handles the scheduling of the execution and the management of resources. For HyP3, we use the AWS Elastic Container Service (ECS) inside of Batch to run the image processing workflows within Docker containers that are unique to each job type. Upon successful processing, the job results are uploaded to S3. The encapsulation of all processing-specific code within Docker containers with a simple interface is an important piece of the HyP3 architecture and is discussed in further detail in the Plugin Development section below. All processing workflows are run on EC2-spot instance to reduce the cost of image processing.

Reviewer 2: ```suggestion
The workflow begins by inspecting the job definition for the job type. This job type will be used to determine which and how many image processing workflows are run and to allocate job-appropriate computing resources. Once the memory and compute resources requirements are determined, the workflow submits the job processing via AWS Batch, which handles the scheduling of the execution and the management of resources. For HyP3, we use the AWS Elastic Container Service (ECS) inside of Batch to run the image processing workflows within Docker containers that are unique to each job type. Upon successful processing, the job results are uploaded to S3. The encapsulation of all processing-specific code within Docker containers with a simple interface is an important piece of the HyP3 architecture and is discussed in further detail in the Plugin Development section below. All processing workflows are run on EC2-spot instance to reduce the cost of image processing.
```

Reviewer 2: The sentence `With this setup, a HyP3 deployment can contain an arbitrary number of different image processing workflows` was a little confusing to me. Are you talking about jobs like `water_map`, that first do RTC then the water map steps?

Reviewer: I *think* it just means that HyP3 can contain any number of plugins? I think we should just define the term "plugin" early in the architecture section, and then use that term throughout the rest of the section. The term "image processing workflows" is too wordy and vague for my taste.

Reviewer 3: We also aren't restricted to "image processing" as the scope for HyP3 plugins, although we can loosely lump everything we currently do under that umbrella.

Side note: I'm writing a proposal to do model simulation with hyp3; there will be nothing image-y about it.

> The workflow begins by inspecting the job definition for the job type. This job type will be used to determine which image processing workflow is run and what computing resources the job needs. With this setup, a HyP3 deployment can contain an arbitrary number of different image processing workflows. Once the memory and compute resources requirements are determined, the workflow submits the job processing via AWS Batch, which handles the scheduling of the execution and the management of resources. For HyP3, we use the AWS Elastic Container Service (ECS) inside of Batch to run the image processing workflows within Docker containers that are unique to each job type. Upon successful processing, the job results are uploaded to S3. The encapsulation of all processing-specific code within Docker containers with a simple interface is an important piece of the HyP3 architecture and is discussed in further detail in the Plugin Development section below. All processing workflows are run on EC2-spot instance to reduce the cost of image processing.

Reviewer 2: Minor suggestions
```suggestion
The workflow begins by inspecting the job definition for the job type. This job type will be used to determine which image processing workflow is run and what computing resources the job needs. With this setup, a HyP3 deployment can contain an arbitrary number of different image processing workflows. Once the memory and compute resources requirements are determined, the workflow submits the job processing via AWS Batch, which handles the scheduling of the execution and the management of resources. For HyP3, we use the AWS Elastic Container Service (ECS) inside of Batch to run the image processing workflows within Docker containers unique to each job type. Upon successful processing, the job results are uploaded to S3. Encapsulating all processing-specific code within Docker containers with a simple interface is an important piece of the HyP3 architecture and is discussed in further detail in the Plugin Development section below. All processing workflows are run on EC2 Spot Instances to reduce the cost of image processing.
```
